### PR TITLE
Added functionality for client SSL certificates to be passed to Rails

### PR DESCRIPTION
**Pull request by rhomeister**
This commit makes it possible to perform client authentication via SSL certificates. The
client provides an SSL certificate which is passed to Rails. Based on custom computation,
controllers can decide whether to grant access.

This commit changes the following:

- It adds a new option `nginx_pass_ssl_client_cert`. If true, nginx is configured to pass
  the SSL certificate of the client to Rails. Rails receives these as
  `request.headers['X-Client-Cert']` or `request.headers['rack.session']['X-Client-Cert']`
- It enables local variables to be passed via the `template` method.
- If `nginx_pass_ssl_client_cert` is true, an additional server directive is generated for
  port 444, which accepts SSL client certificate authentication.

If this pull request is accepted. I will add additional documentation and information on how to create controllers for validating the SSL certificate of the client. The main benefit of passing SSL certificates to Rails, instead of configuring these in nginx, is that no nginx configuration has to be changed when adding or revoking SSL certificates. Moreover, different controllers can allow different SSL sets of certificates, without creating additional server or location directives in nginx conf.

### CODE REVIEW

**Comment by rhomeister** [comment]
Note that the biggest change is a refactor of the nginx_conf.erb. SSL client authentication operates on a different port. Since the server directives of the main port (443) and the SSL client authentication port are almost identical, I dried the code a little by rendering the server directive template twice.

File: lib/capistrano/tasks/nginx.rake [hunk 1]
```diff
@@ -13,6 +13,9 @@ namespace :load do
     # ssl options
     set :nginx_location, '/etc/nginx'
     set :nginx_use_ssl, false
+    set :nginx_use_spdy, false
```
Comment (bruno-) [inline lib/capistrano/tasks/nginx.rake]: You wanted to include this line in this pull request? I think the setting makes sense but maybe add it in a separate commit. Feel free to push it straight to master.

File: lib/generators/capistrano/unicorn_nginx/templates/nginx_conf.erb [hunk 2]
```diff
@@ -16,71 +16,10 @@ server {
...
-
-  if (-f $document_root/system/maintenance.html) {
-    return 503;
-  }
-
-}
+<% if fetch(:nginx_pass_ssl_client_cert) %>
+<% # render the server directive with SSL client certificate authentication enabled on port 444 %>
+<%= template_to_s("_default_server_directive.erb", ssl_port: 444, nginx_pass_ssl_client_cert: true).to_s %>
+<% end %>
```
Comment (bruno-) [inline lib/generators/capistrano/unicorn_nginx/templates/nginx_conf.erb]: I know this is mostly aesthetic, but can you please make your editor use newlines for the EOF too. More thorough [explanation here](http://robots.thoughtbot.com/no-newline-at-end-of-file).

**Comment by bruno-** [comment]
Hi @rhomeister,
the feature blows my mind since I'm not sure I understand what to use it for. But I'm okay this gets merged if you use this.

I left 2 minor comments in the code. It would be great if you fixed those.
Also, can you please update the `CHANGELOG` file?

Other than this, merge when you're ready.

**Comment by bruno-** [comment]
Hi @rhomeister,
I've also added you as a gem owner in #49 :tada: :tada:
That way you can make a new release to rubygems.org when this is merged.

**Comment by rhomeister** [comment]
@bruno- Done. I've also added a lot more documentation and explanation about this feature to the wiki.

**Comment by rhomeister** [comment]
Released as 3.3.0

**Pull request merged**